File: android/feature/catalog/src/main/kotlin/SeasonWall.kt
package catalog

/**
 * One season's plate on a show's wall.
 *
 * [posterKey] is the season's own artwork key, derived from the show's —
 * `<showKey>-s<n>`, the same rule the index's fetch already downloads
 * artwork under (`mlib_spec::package::season_poster_key`). It is `null`
 * for a division with no season number — "Episodes", specials — and for
 * a show that has no poster key of its own to build one from; either way
 * the plate falls back to the show's poster, the same way any other
 * missing poster does.
 *
 * [watched] is true once every episode under it is — the only sense in
 * which a season is watched, ported from `shelf-view.js`'s `seasonGrid`.
 */
data class SeasonPlate(
    val title: String,
    val caption: String,
    val posterKey: String?,
    val division: Division,
    val watched: Boolean,
)

/**
 * The seasons wall for [collection], or `null` when there is nothing to
 * wall. [watchedIds] is this viewer's finished sets, for [SeasonPlate.watched].
 *
 * A course is never walled: it drills into chapters, not seasons, and
 * those are shown as the nested tree they already are. Nor is a show with
 * one season — a wall of one plate is a tap that says nothing a direct
 * list of its episodes did not already say, so that case is left to fall
 * through to the flat list too.
 */
fun seasonPlatesOf(
    collection: Entry.Collection,
    watchedIds: Set<String> = emptySet(),
): List<SeasonPlate>? {
    if (collection.kind != CollectionKind.SHOW || collection.divisions.size <= 1) return null
    return collection.divisions.map { division ->
        val items = division.walk().flatMap { it.items }.toList()
        SeasonPlate(
            title = division.title,
            caption = "${items.size} ${plural(items.size, "episode")}",
            posterKey =
                collection.posterKey?.let { showKey ->
                    division.season?.let { season -> seasonPosterKey(showKey, season) }
                },
            division = division,
            watched = items.isNotEmpty() && items.all { it.setId in watchedIds },
        )
    }
}

/**
 * The key a season's artwork is held under, beside its show's own poster
 * key — mirrors `mlib_spec::package::season_poster_key` on the Rust side,
 * which is what the fetch that downloads this artwork names it with.
 */
fun seasonPosterKey(
    showPosterKey: String,
    season: Int,
): String = "$showPosterKey-s$season"

private fun plural(
    count: Int,
    word: String,
): String = if (count == 1) word else "${word}s"

File: web/public/lib/season-wall.js
/**
 * What a show's page opens on: one plate per season.
 *
 * A show of several seasons is a wall of them, each with its own artwork,
 * and opening one lists its episodes — the way a course's folders are doors.
 * A show of one season goes straight to its episodes: a wall of one plate is
 * a click that says nothing.
 *
 * No DOM here; the plate itself is `shelf-view.js`'s card, so a season looks
 * like everything else on a wall.
 */

import { countOf } from "./format.js";

/** Whether a show's page is a wall of seasons rather than its episodes. */
export function hasSeasonWall(collection) {
  return collection.divisions.length > 1;
}

/**
 * The plate for one season.
 *
 * The season's own artwork when TMDB has one, else the show's — every
 * episode carries both, so the first is enough to ask. A season with neither
 * falls to initials, as any plate does.
 */
export function seasonPlate(division) {
  const first = division.items[0] ?? null;
  return {
    name: division.title,
    meta: countOf(division.items.length, "episode"),
    poster: first?.seasonPoster ?? first?.poster ?? null,
  };
}

/** The season a page's trail names, by the title its plate carried. */
export function seasonNamed(collection, title) {
  return collection.divisions.find((division) => division.title === title) ?? null;
}
